### PR TITLE
lis2ds12 accelerometer alignment

### PR DESCRIPTION
**Pull request by avisconti**
Improvements:                                                                                
                                                                                
0. Aligned stmesc API to latest DT facilities                                                                               
1. Make use of latest stmemsc module lis2ds12_STdC APIs                                                                                
2. Add multi-instance support                                                                                
3. Move range and odr into DT                                                                                
                                                                                
Tested with x_nucleo_iks01a3 shield in SPI mode

### CODE REVIEW

File: drivers/sensor/lis2ds12/lis2ds12_spi.c [hunk 1]
```diff
@@ -94,68 +90,24 @@ static int lis2ds12_raw_write(struct lis2ds12_data *data, uint8_t reg_addr,
...
-
-static const struct lis2ds12_transfer_function lis2ds12_spi_transfer_fn = {
-	.read_data = lis2ds12_spi_read_data,
-	.write_data = lis2ds12_spi_write_data,
-	.read_reg  = lis2ds12_spi_read_reg,
-	.write_reg  = lis2ds12_spi_write_reg,
-	.update_reg = lis2ds12_spi_update_reg,
+stmdev_ctx_t lis2ds12_spi_ctx = {
+	.read_reg = (stmdev_read_ptr) lis2ds12_spi_read,
+	.write_reg = (stmdev_write_ptr) lis2ds12_spi_write,
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12_spi.c]: nit: extra space after `)`

File: drivers/sensor/lis2ds12/lis2ds12.c [hunk 2]
```diff
@@ -267,32 +240,121 @@ static int lis2ds12_init(const struct device *dev)
...
+			.read_reg =					\
+			   (stmdev_read_ptr) stmemsc_spi_read,		\
+			.write_reg =					\
+			   (stmdev_write_ptr) stmemsc_spi_write,	\
+			.handle =					\
+			   (void *)&lis2ds12_config_##inst.stmemsc_cfg,	\
+		},							\
+		.stmemsc_cfg.spi = {					\
+			.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),	\
+			.spi_cfg = SPI_CONFIG_DT_INST(inst,		\
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12.c:372]: `SPI_DT_SPEC_INST_GET` should be used

File: drivers/sensor/lis2ds12/lis2ds12.c [hunk 3]
```diff
@@ -267,32 +240,121 @@ static int lis2ds12_init(const struct device *dev)
...
+			.read_reg =					\
+			   (stmdev_read_ptr) stmemsc_i2c_read,		\
+			.write_reg =					\
+			   (stmdev_write_ptr) stmemsc_i2c_write,	\
+			.handle =					\
+			   (void *)&lis2ds12_config_##inst.stmemsc_cfg,	\
+		},							\
+		.stmemsc_cfg.i2c = {					\
+			.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),	\
+			.i2c_slv_addr = DT_INST_REG_ADDR(inst),		\
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12.c:399]: `I2C_DT_SPEC_INST_GET`

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c [hunk 4]
```diff
@@ -28,8 +24,8 @@ static void lis2ds12_gpio_callback(const struct device *dev,
 
 	ARG_UNUSED(pins);
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_DISABLE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12_trigger.c:27]: this function should be checked for errors

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c [hunk 5]
```diff
@@ -60,8 +55,8 @@ static void lis2ds12_handle_int(const struct device *dev)
 		lis2ds12_handle_drdy_int(dev);
 	}
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12_trigger.c:58]: ditto

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c [hunk 6]
```diff
@@ -151,8 +160,8 @@ int lis2ds12_trigger_init(const struct device *dev)
 	data->work.handler = lis2ds12_work_cb;
 #endif
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12_trigger.c:163]: ditto

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c [hunk 7]
```diff
@@ -162,14 +171,19 @@ int lis2ds12_trigger_set(const struct device *dev,
...
 	__ASSERT_NO_MSG(trig->type == SENSOR_TRIG_DATA_READY);
 
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_DISABLE);
+	if (cfg->gpio_int.port == NULL) {
+		LOG_ERR("trigger_set is not supported");
+		return -ENOTSUP;
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12_trigger.c:185]: ditto

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c [hunk 8]
```diff
@@ -183,8 +197,8 @@ int lis2ds12_trigger_set(const struct device *dev,
 	data->data_ready_trigger = *trig;
 
 	lis2ds12_init_interrupt(dev);
-	gpio_pin_interrupt_configure(data->gpio, cfg->irq_pin,
-				     GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
```
Comment (gmarull) [inline drivers/sensor/lis2ds12/lis2ds12_trigger.c:200]: ditto

**Review by gmarull — CHANGES_REQUESTED** [review]
Changes lgtm, some minor comments that I think should be addressed

File: drivers/sensor/lis2ds12/lis2ds12.c (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12.c:372]: It seems fine, as it is a code semplification. But I have to change it into stmemsc_cfg_spi structure, which is common to many ST sensor now. So, I have to do it as a preliminary commit for those sensors first, and I will do it in this PR itself. Sounds good?

File: drivers/sensor/lis2ds12/lis2ds12.c [hunk 9]
```diff
@@ -41,98 +41,65 @@ static struct lis2ds12_config lis2ds12_config = {
...
-
-static int lis2ds12_accel_range_to_fs_val(int32_t range)
-{
-	size_t i;
-
-	for (i = 0; i < ARRAY_SIZE(lis2ds12_accel_fs_map); i++) {
-		if (range == lis2ds12_accel_fs_map[i]) {
-			return i;
-		}
+	val =  LIS2DS12_HR_ODR_TO_REG(odr);
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12.c]: ack

File: drivers/sensor/lis2ds12/lis2ds12.c [hunk 10]
```diff
@@ -41,98 +41,65 @@ static struct lis2ds12_config lis2ds12_config = {
...
+	if (val > LIS2DS12_XL_ODR_800Hz_HR) {
+		LOG_ERR("ODR too high");
+		return -ENOTSUP;
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12.c:92]: ack

File: drivers/sensor/lis2ds12/lis2ds12.c [hunk 11]
```diff
@@ -41,98 +41,65 @@ static struct lis2ds12_config lis2ds12_config = {
...
-
-	if (data->hw_tf->update_reg(data,
-				    LIS2DS12_REG_CTRL1,
-				    LIS2DS12_MASK_CTRL1_FS,
-				    fs << LIS2DS12_SHIFT_CTRL1_FS) < 0) {
-		LOG_DBG("failed to set accelerometer full-scale");
-		return -EIO;
+	switch (range) {
+	default:
+	case 2:
```
[the remark this answers is not in the document]
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12.c:107]: ack

File: drivers/sensor/lis2ds12/lis2ds12_spi.c (hunk 1 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_spi.c]: ack

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 4 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:27]: But the callback returns void in this case. Do you want me to raise a LOG_ERR here?

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 5 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:58]: same here. LOG_ERR?

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 6 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:163]: ack

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 7 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:185]: ack

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 8 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:200]: ack

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 4 above)
Comment (gmarull) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:27]: Maybe an assert then?

File: drivers/sensor/lis2ds12/lis2ds12.c (hunk 2 above)
Comment (gmarull) [reply drivers/sensor/lis2ds12/lis2ds12.c:372]: It’s ok, just raised as it’s always better to use existing facilities when available.

File: drivers/sensor/lis2ds12/lis2ds12_trigger.c (hunk 4 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12_trigger.c:27]: sure

File: drivers/sensor/lis2ds12/lis2ds12.c (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/lis2ds12/lis2ds12.c]: > It’s ok, just raised as it’s always better to use existing facilities when available.

Yes, correct. THanks for pointing this out. The stmemsc common APIs are now quite simpler.

**Review by gmarull — APPROVED** [review]

**Comment by jefffhaynes** [comment]
@MaureenHelm @mbolivar-nordic I could really use this merge.  Any chance of getting some reviews?

**Comment by avisconti** [comment]
> @MaureenHelm @mbolivar-nordic I could really use this merge. Any chance of getting some reviews?

@jefffhaynes I think this PR was blocked by v2.7.0 release.

**Review by MaureenHelm — COMMENTED** [review]
LGTM.

@avisconti will you please rebase? I'd like to see a fresh CI run on this PR before merging.

**Comment by avisconti** [comment]
Still running. Mmmmh...

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**